switch from states mapping to two state variables for the genesis and…

alysiahuggins opened the pull request:
… finalized state

Closes #1938 
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
- removes states mapping
- added global variable for genesisState and finalizedState
- WIP
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->

### This PR does not:
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

### Key places to review:
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->

philippecamacho commented:
Made some progress fixing the compilation errors but not there yet:
* Created a new [branch](https://github.com/EspressoSystems/HotShot/tree/lc-contract-updates) of HotShot (fork of tag = "0.5.72") and added the `StakeState` struct.
* Have the espresso-sequencer repository point to this branch
* Fix several compilation errors.

Right now, for some reason the espresso-types crates does not compile.

alysiahuggins commented:
feature branch, `lc-contract-updates` have been added to the following repos so that we can support the new types introduced by this ticket:
- https://github.com/EspressoSystems/HotShot
- https://github.com/EspressoSystems/hotshot-builder-core
- https://github.com/EspressoSystems/marketplace-builder-core
- https://github.com/EspressoSystems/hotshot-events-service
- https://github.com/EspressoSystems/hotshot-query-service
https://github.com/EspressoSystems/espresso-sequencer/commit/d391f7e16953566b229245ec65713196616e6650

alysiahuggins commented:
- removed fields from the `LightClientState` struct so it now looks like this 
```
struct LightClientState {
        uint64 viewNum;
        uint64 blockHeight;
        BN254.ScalarField blockCommRoot;
    }
```
- updated related rust code such as the lightclient `adapter` and `MockGenesis` action in the `diff-tests`, to continue, one would have to also update the `MockConsecutiveFinalizedStates`, `MockSkipBlocks`, `MockMissEndingBlock` etc


https://github.com/EspressoSystems/espresso-sequencer/pull/1939/commits/3a27dda9fce8ab076fbd629ea9b4e327e9f39c2c
this git commit ignored verification checks so that anyone can pick up from where i left off.

alxiong commented on contracts/src/LightClient.sol:
```diff
@@ -91,20 +94,22 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
+    /// @param stakeTableBlsKeyComm The commitment to the BlsVerKey column of the stake table
+    /// @param stakeTableSchnorrKeyComm The commitment to the SchnorrVerKey column of the table
+    /// @param stakeTableAmountComm The commitment to the stake amount column of the stake table
+    struct StakeState {
+        uint256 threshold;
         BN254.ScalarField stakeTableBlsKeyComm;
         BN254.ScalarField stakeTableSchnorrKeyComm;
         BN254.ScalarField stakeTableAmountComm;
```
naming nit:
```suggestion
    struct StakeTableState {
        uint256 threshold;
        BN254.ScalarField blsKeyComm;
        BN254.ScalarField schnorrKeyComm;
        BN254.ScalarField amountComm;
```

reason: 
- `StakeTableState` is clearer imo
- once we have the enclosing struct name to have the "stake table" prefix, we don't need it in the names of its fields, because `stakeTableState.blsKeyComm` already provides a clear context.

alxiong replied:
btw, @alysiahuggins i'm addressing my own comments locally. just FYI to avoid redundant work

alxiong replied:
done in [f7c5afa](https://github.com/EspressoSystems/espresso-sequencer/pull/1939/commits/f7c5afa0b52200c2e9ac8573ef6ca92c7aed7471)

alxiong commented on contracts/src/LightClient.sol:
```diff
@@ -62,6 +56,15 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
     /// @notice mapping to store light client states in order to simplify upgrades
     mapping(uint32 index => LightClientState value) public states;
```
```suggestion
```

we should remove this by now?

alysiahuggins replied:
ah yes thank you, done here https://github.com/EspressoSystems/espresso-sequencer/commit/3f4417012c9b75e967b24de3ae35b02b9c7f504b

alxiong commented on contracts/src/LightClient.sol:
```diff
@@ -199,34 +203,38 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
-        bytes32 initStakeTableComm = computeStakeTableComm(_genesis);
+        bytes32 initStakeTableComm = computeStakeTableComm(_genesisStakeState);
         votingStakeTableCommitment = initStakeTableComm;
-        votingThreshold = _genesis.threshold;
+        votingThreshold = _genesisStakeState.threshold;
         frozenStakeTableCommitment = initStakeTableComm;
-        frozenThreshold = _genesis.threshold;
+        frozenThreshold = _genesisStakeState.threshold;
```
imo, we should remove all of these: 
```solidity
bytes32 public votingStakeTableCommitment;
uint256 public votingThreshold;
bytes32 public frozenStakeTableCommitment;
uint256 public frozenThreshold;
mapping(uint32 index => LightClientState value) public states;

function computeStakeTableComm(LightClientState memory state) returns (bytes32);
```

i roughly explain the "frozen/voting stake table" here: https://www.notion.so/espressosys/Light-Client-Contract-Details-6c396f7824e34356bdcda710015291f9?pvs=4#57affaec4dbf4508a27b169255fde936

TL;DR is `votingXX` is the stake table snapshot/version that we will be using when validating quorum signatures (or SNARK that proves a valid QC)

but since we are not changing them anyway, just remove them for now and use `genesisState` you introduced.

alxiong replied:
addressed in [c63635e](https://github.com/EspressoSystems/espresso-sequencer/pull/1939/commits/c63635e04eb70c651e8bd2d3bf7295e1ee5f9b80)

alxiong commented on contracts/src/LightClient.sol:
```diff
@@ -282,12 +285,17 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
     /// @dev Simple getter function for the genesis state
     function getGenesisState() public view virtual returns (LightClientState memory) {
-        return states[genesisState];
+        return genesisState;
+    }
+
+    /// @dev Simple getter function for the genesis stake state
+    function getGenesisStakeState() public view virtual returns (StakeState memory) {
+        return genesisStakeState;
     }
 
     /// @dev Simple getter function for the finalized state
     function getFinalizedState() public view virtual returns (LightClientState memory) {
-        return states[finalizedState];
+        return finalizedState;
     }
```
remove all of these helpers because these states are trivial to read (and come with default getter)

alysiahuggins replied:
ah yes, it was used when we had the mapping, updated here https://github.com/EspressoSystems/espresso-sequencer/commit/56b88cd34eeab65ce00ba3e8a954accfc850937f

alxiong commented on hotshot-state-prover/src/mock_ledger.rs:
```diff
@@ -347,21 +362,16 @@ impl MockLedger {
 
     /// Returns the (bytes32 votingStakeTableComm, bytes32 frozenStakeTableComm) used in contract
-    pub fn get_stake_table_comms(&self) -> (H256, H256) {
+    pub fn get_stake_table_comms(&self) -> (H256, H256, ParsedStakeState) {
```
is there a reason why we still return the first two commitments?
maybe renamed to `get_stake_table_state(&self) -> ParsedStakeTableState`

alxiong commented on hotshot-state-prover/src/mock_ledger.rs:
```diff
@@ -99,14 +102,14 @@ impl MockLedger {
             block_comm_root,
-            fee_ledger_comm,
-            stake_table_comm: st.commitment(SnapshotVersion::LastEpochStart).unwrap(),
+            // fee_ledger_comm,
+            // stake_table_comm: st.commitment(SnapshotVersion::LastEpochStart).unwrap(),
```
just a reminder to remove all commented lines later. no need to address it now.

alxiong commented on hotshot-state-prover/src/circuit.rs:
needs to update the code comments (and docs if any) to reflect the changes.

alxiong commented on hotshot-state-prover/src/service.rs:
```diff
@@ -200,6 +200,35 @@ pub fn light_client_genesis_from_stake_table(
+}
+
+#[inline]
+pub fn light_client_genesis_stake_from_stake_table(
+    st: StakeTable<BLSPubKey, StateVerKey, CircuitField>,
+) -> anyhow::Result<ParsedStakeState> {
```
shouldn't we use `StakeTable.into()` I thought `From<hotshot_stake_table::StakeTable> for ParsedStakeTableState` is implemented in `adapters/light_client.rs`?

if not, we should add it there; and we can remove this function.

alxiong reviewed (commented):
I've looked at `adapter/*`, `diff-test` , `hotshot-state-prover`
they all look very good to me. Alysia is doing great work. 👏

alxiong commented:
@mrain can you also take a look at Alysia's changes to the `circuit.rs`. The changes look good to me at least. 

for reference, we are simplifying light client state as specified in https://github.com/EspressoSystems/espresso-sequencer/issues/1938

thanks!

mrain commented:
> @mrain can you also take a look at Alysia's changes to the `circuit.rs`. The changes look good to me at least.
> 
> for reference, we are simplifying light client state as specified in #1938
> 
> thanks!

LGTM. Changes are simple enough.

philippecamacho commented on contracts/rust/diff-test/src/main.rs:
```diff
@@ -503,9 +512,17 @@ fn main() {
+                    schnorr_key_comm: field_to_u256(stt.stake_table_schnorr_key_comm),
+                    amount_comm: field_to_u256(stt.stake_table_amount_comm),
+                };
+                stake_stakes.push(parsed_stake_state.into());
```
Looks like `stake_stakes` vector is not returned. Why is that?

alxiong replied:
this will be removed in my future commit of mine

alxiong replied:
removed in [af51261](https://github.com/EspressoSystems/espresso-sequencer/pull/1986/commits/af5126122179c8c58999c32b3a084b1bd8e4a2c0)

alxiong commented on hotshot-state-prover/src/circuit.rs:
```diff
@@ -68,10 +68,6 @@ impl LightClientStateVar {
                 circuit.create_public_variable(view_number_f)?,
                 circuit.create_public_variable(block_height_f)?,
                 circuit.create_public_variable(state.block_comm_root)?,
-                circuit.create_public_variable(state.fee_ledger_comm)?,
-                circuit.create_public_variable(state.stake_table_comm.0)?,
-                circuit.create_public_variable(state.stake_table_comm.1)?,
-                circuit.create_public_variable(state.stake_table_comm.2)?,
```
ah this is wrong actually. 🤦 
(i wasted some time assuming our Public input is reduced to 4) 

we should still keep the stake table commitments in the public inputs for security/integrity!!

but we can safely remove `fee_ledger_comm`, so PI.length is 7 now.

cc @mrain @alysiahuggins

alxiong replied:
ah, I need to fix `HotShot` side as well for the struct declaration.

alxiong replied:
fixed in [ab81b0b](https://github.com/EspressoSystems/espresso-sequencer/pull/1986/commits/ab81b0bad7497dbcedc4434cd1881cb6891d7c58)

alxiong commented on hotshot-state-prover/src/circuit.rs:
```diff
@@ -292,32 +283,32 @@ where
         .map(|var| var.stake_amount)
         .collect::<Vec<_>>();
-    let stake_amount_comm = RescueNativeGadget::<F>::rescue_sponge_with_padding(
+    let _stake_amount_comm = RescueNativeGadget::<F>::rescue_sponge_with_padding(
         &mut circuit,
         &stake_amount_preimage_vars,
         1,
     )?[0];
-    circuit.enforce_equal(
-        stake_amount_comm,
-        lightclient_state_pub_var
-            .stake_table_comm()
-            .stake_amount_comm,
-    )?;
```
realized that these are wrong, we need these to ensure security. 

sry that I miss it during my last review.

alxiong replied:
fixed in [ab81b0b](https://github.com/EspressoSystems/espresso-sequencer/pull/1986/commits/ab81b0bad7497dbcedc4434cd1881cb6891d7c58)
